transform/interrupt.go

```diff
@@ -73,25 +73,50 @@ func LowerInterrupts(mod llvm.Module, sizeLevel int) []error {
 	// Create a function type with the signature of an interrupt handler.
 	fnType := llvm.FunctionType(ctx.VoidType(), nil, false)
 
-	// Collect a slice of interrupt handle objects. The fact that they still
+	// Collect a map of interrupt handle objects. The fact that they still
 	// exist in the IR indicates that they could not be optimized away,
 	// therefore we need to make real interrupt handlers for them.
-	var handlers []llvm.Value
+	handleMap := map[int][]llvm.Value{}
 	handleType := mod.GetTypeByName("runtime/interrupt.handle")
 	if !handleType.IsNil() {
 		handlePtrType := llvm.PointerType(handleType, 0)
 		for global := mod.FirstGlobal(); !global.IsNil(); global = llvm.NextGlobal(global) {
 			if global.Type() != handlePtrType {
 				continue
 			}
-			handlers = append(handlers, global)
+
+			// Get the interrupt number from the initializer
+			initializer := global.Initializer()
+			num := int(llvm.ConstExtractValue(initializer, []uint32{1, 0}).SExtValue())
+			pkg := packageFromInterruptHandle(global)
+
+			handles, exists := handleMap[num]
+
+			// If there is an existing interrupt handler, ensure it is in the same package
+			// as the new one.  This is to prevent any assumptions in code that a single
+			// compiler pass can see all packages to chain interrupt handlers. When packages are
+			// compiled to separate object files, the linker should spot the duplicate symbols
+			// for the wrapper function, failing the build.
+			if exists && packageFromInterruptHandle(handles[0]) != pkg {
+				errs = append(errs, errorAt(global,
+					fmt.Sprintf("handlers for interrupt %d (%s) in multiple packages: %s and %s",
+						num, handlerNames[int64(num)], pkg, packageFromInterruptHandle(handles[0]))))
+				continue
+			}
+
+			handleMap[num] = append(handles, global)
 		}
 	}
 
-	// Iterate over all handler objects, replacing their ptrtoint uses with a
+	// Iterate over all handle objects, replacing their ptrtoint uses with a
 	// real interrupt ID and creating an interrupt handler for them.
-	for _, global := range handlers {
-		initializer := global.Initializer()
+	for _, handles := range handleMap {
```

Comment:
This introduces non-deterministic behavior in the compiler because the iteration over the map is done in unspecified order.

Example:

```
$ tinygo build -o test.hex -target=microbit examples/serial && md5sum test.hex
59c5d997db197b0e4707f7541935be4f  test.hex

$ tinygo build -o test.hex -target=microbit examples/serial && md5sum test.hex
ab431d70b7686fe593c8fbd7073ad885  test.hex
```

Reply:
Thanks - yes, that definitely needs fixing.  I'll work on a fix.

Reply:
I've pushed a fix to sort the handlers in interrupt number order.  Builds are now repeatable:
```shell
$ tinygo build -o test.hex -target=microbit examples/serial && md5sum test.hex
d988eef1e3bc84fc8def75acb677470c  test.hex
$ tinygo build -o test.hex -target=microbit examples/serial && md5sum test.hex
d988eef1e3bc84fc8def75acb677470c  test.hex
```